src/drivers/hermes/hermes.c

```diff
@@ -85,11 +192,43 @@ static ssize_t hermes_read(struct file *filp, char __user *buf, size_t len,
 static ssize_t hermes_write(struct file *filp, const char __user *buf,
 		size_t len, loff_t *off)
 {
-	return len;
+	int ret = 0;
+	struct hermes_cmd_req req;
+
+	if (len != sizeof(req)) {
+		ret = -ENOSPC;
+		goto out;
+	}
+
+	if (copy_from_user(&req, buf, len)) {
+		ret = -EFAULT;
+		goto out;
+	}
+
+	switch(req.opcode) {
```

Comment:
This *really* looks like it should be an IOCTL, not a write.

Reply:
@lsgunth would moving this to an ioctl() impact how this driver interacts with io_uring?

Reply:
Well yes, an ioctl won't work with io_uring.... But that doesn't mean we should go abusing writes calls just to work with io_uring.

This design doesn't really get much benefit from io_uring seeing everything must be synchronous and linked with each other.

Another option might be throwing out the XDMA driver completely (it's really bad anyway)... Then make it so you do a write() followed by a read() on a char device. The write triggers the XDMA source data transfer, the read() at offset 0 triggers the algorithm then the read back of the result data. I actually like that a lot better than the mess that's currently being created.

Reply:
I'm not sure I understand your suggestion. As we discussed before, we can't drop the whole XDMA engine as we don't have the resources to implement an alternative DMA engine. So in terms of dropping just the XDMA driver, I don't understand how your suggestion is fundamentally different. The XDMA's write() does essentially what you said, with the difference that you have to explicitly select a channel, but that's easy to fix.

The read() would be quite similar as well, with the addition of triggering the eBPF engine first. Also, how would you suggest specifying which program/data slots to use with this design?

Reply:
The suggestion of dropping the xdma driver is so userspace doesn't have to interact with two different drivers to interact with one piece of hardware.

> The read() would be quite similar as well, with the addition of triggering the eBPF engine first. Also, how would you suggest specifying which program/data slots to use with this design?

The kernel would be responsible for assigning and freeing channels and slots... as it should be. The current design is also very broken from a security perspective as it doesn't prevent userspace from just DMAing to any random slot or address they want.

I still vote for ditching the xdma driver. It doesn't do what we want and is a much larger mess than we need. It also seems really poorly written -- a driver doing what it does shouldn't be more than a single file and a several hundred lines.

Reply:
The idea in this PR wasn't for userspace to interact with two drivers, only with the hermes driver through the hermes char device. The fact that the hermes driver makes calls into the xdma driver are hidden from the user. Anyway, there's the option of making it a single driver as you commented previously.

Regarding security, the idea was to eventually get rid of the /dev/xdma* devices, so that everything has to go through /dev/hermesX.

Yes, the XDMA driver is quite large. But there are several parts of it that we aren't using and that we can remove.